### PR TITLE
GNU hello recipe

### PR DESCRIPTION
**Pull request by theclue**
Recipe for GNU 'hello world' demo program.

Ported for learning purposes, submitted for getting feedback about errors, missing or misconfigurations in recipe

Had difficulties to understand if the REQ is about the binary gettext or the devel library. Added both, but it's probably wrong.

tnx for support!

### CODE REVIEW

File: app-misc/hello/hello-2.10.recipe [hunk 1]
```diff
@@ -0,0 +1,49 @@
+SUMMARY="GNU Hello program"
+DESCRIPTION="The GNU Hello program produces a familiar, friendly greeting.\
+Yes, this is another implementation of the classic program that prints \"Hello, world!\" when you run it.\
+However, unlike the minimal version often seen, GNU Hello processes its argument list to modify its behavior,\
```
Comment (fbrosson) [inline app-misc/hello/hello-2.10.recipe]: Suggestion:
```diff
-SUMMARY="GNU Hello program"
+SUMMARY="An implementation of the \"Hello World!\" program"
-DESCRIPTION="The GNU Hello program produces a familiar, friendly greeting.\
-Yes, this is another implementation of the classic program that prints \"Hello, world!\" when you run it.\
-However, unlike the minimal version often seen, GNU Hello processes its argument list to modify its behavior,\
-supports greetings in many languages, and so on.\
-The primary purpose of GNU Hello is to demonstrate how to write other programs that do these things;\
+DESCRIPTION="The GNU Hello program produces a familiar, friendly greeting. \
+Yes, this is another implementation of the classic program that prints \
+\"Hello, world!\" when you run it. However, unlike the minimal version often \
+seen, GNU Hello processes its argument list to modify its behavior, supports \
+greetings in many languages, and so on. \
+The primary purpose of GNU Hello is to demonstrate how to write other programs \
+that do these things; \
 it serves as a model for GNU coding standards and GNU maintainer practices."
```

File: app-misc/hello/hello-2.10.recipe [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+supports greetings in many languages, and so on.\
+The primary purpose of GNU Hello is to demonstrate how to write other programs that do these things;\
+it serves as a model for GNU coding standards and GNU maintainer practices."
+HOMEPAGE="https://www.gnu.org/software/hello/"
+COPYRIGHT="1992-2014 Free Software Foundation, Inc."
+LICENSE="GNU GPL v3"
+REVISION="2"
+SOURCE_URI="http://ftp.gnu.org/gnu/hello/hello-$portVersion.tar.gz"
```
Comment (fbrosson) [inline app-misc/hello/hello-2.10.recipe]:
```diff
-REVISION="2"
-SOURCE_URI="http://ftp.gnu.org/gnu/hello/hello-$portVersion.tar.gz"
+REVISION="1"
+SOURCE_URI="https://ftpmirror.gnu.org/hello/hello-$portVersion.tar.gz
+	https://ftp.gnu.org/gnu/hello/hello-$portVersion.tar.gz"
```

File: app-misc/hello/hello-2.10.recipe [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+CHECKSUM_SHA256="31e066137a962676e89f69d1b65382de95a7ef7d914b8cb956f41ea72e0f516b"
+
+ARCHITECTURES="x86 x86_gcc2 x86_64"
```
Comment (fbrosson) [inline app-misc/hello/hello-2.10.recipe]:
```diff
-ARCHITECTURES="x86 x86_gcc2 x86_64"
+ARCHITECTURES="x86_gcc2 x86 x86_64"
```
(Most of our recipes list them in this order, so we try to stick to it :)

File: app-misc/hello/hello-2.10.recipe [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+BUILD_REQUIRES="
+	haiku_devel
+	devel:libgettextlib >= 0.18.1
+	
+	"
+BUILD_PREREQUIRES="
+	cmd:autoconf
+	cmd:gcc
+	cmd:make
```
Comment (fbrosson) [inline app-misc/hello/hello-2.10.recipe]: * No need for the empty line after `devel:libgettextlib`
* Sorting `BUILD_PREREQUIRES`

```diff
 	devel:libgettextlib >= 0.18.1
-	
 	"
 BUILD_PREREQUIRES="
 	cmd:autoconf
 	cmd:gcc
-	cmd:make
-	cmd:help2man
 	cmd:gettext
+	cmd:help2man
+	cmd:make
 	"
```

File: app-misc/hello/hello-2.10.recipe [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+BUILD()
+{
+	runConfigure ./configure
+	make
+}
+
+INSTALL()
+{
+	make install
+}
```
Comment (fbrosson) [inline app-misc/hello/hello-2.10.recipe]:
```diff
 INSTALL()
 {
 	make install
+
+	rm "$libDir"/charset.alias
 }
```
(We don't need this file.)

File: app-misc/hello/hello-2.10.recipe (hunk 1 above)
Comment (fbrosson) [reply app-misc/hello/hello-2.10.recipe]: I forgot to say that this is the same text as you had, except that I folded the lines differently, in order to respect the recommended 80 chars limit (per line) in this part of the recipe.

File: app-misc/hello/hello-2.10.recipe [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+ARCHITECTURES="x86 x86_gcc2 x86_64"
+
+PROVIDES="
+	hello = $portVersion
+	cmd:hello = $portVersion
+	"
+REQUIRES="
+	haiku
+	gnulib
```
Comment (fbrosson) [inline app-misc/hello/hello-2.10.recipe]: It looks like the hello program does not need gnulib.
```diff
 REQUIRES="
 	haiku
- 	gnulib
 	lib:libintl
 	"
```

And it does not seem to be required at run time.
Well, maybe this is an optional build dependency. In that case we can add this:
```diff
 BUILD_PREREQUIRES="
 	cmd:autoconf
 	cmd:gcc
 	cmd:gettext
+	cmd:gnulib_tool
 	cmd:help2man
 	cmd:make
 	"
```

Let's see what the other contributors say about all this.

Comment (theclue) [reply app-misc/hello/hello-2.10.recipe]: i've found about this dep into the README.dev. Is there a better way to check for dependencies (and the minimum version)?

**Comment by theclue** [comment]
I've seen that the Travis CI check was failed...did I miss something?

**Comment by Begasus** [comment]
Hi TheClue, thanks for the PR! Line 30 has a "tab" in it @fbrosson already mentioned that the line is not needed, if you make the changes suggested above save them and commit them here it should be ok (for now) ;)

File: app-misc/hello/hello-2.10.recipe [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	lib:libintl
+	"
+
+BUILD_REQUIRES="
+	haiku_devel
+	devel:libgettextlib >= 0.18.1
```
Comment (Begasus) [inline app-misc/hello/hello-2.10.recipe]: This should do also:
```diff
-	devel:libgettextlib >= 0.18.1
+	devel:libintl
```

File: app-misc/hello/hello-2.10.recipe (hunk 6 above)
Comment (Begasus) [reply app-misc/hello/hello-2.10.recipe]: No need for the gnulib_tool :)

File: app-misc/hello/hello-2.10.recipe [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+	
+	"
+BUILD_PREREQUIRES="
+	cmd:autoconf
```
Comment (Begasus) [inline app-misc/hello/hello-2.10.recipe]: no need for autoconf
```diff
-	cmd:autoconf
+	cmd:awk
```

File: app-misc/hello/hello-2.10.recipe [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	cmd:gcc
+	cmd:make
+	cmd:help2man
+	cmd:gettext
```
Comment (Begasus) [inline app-misc/hello/hello-2.10.recipe]: No need for cmd:gettext (provided by lib:linintl)
```diff
-	cmd:gettext
```

**Comment by fbrosson** [comment]
In most cases the contributor who creates the pull request updates it with either new commits or replaces the existing commit(s) (or some of the commits) with one or more commits. Most users prefer to do this by editing their local branch, and then update their remote branch with git (on the command line.) A few users choose to edit their PR using GitHub's web site, but this method is not as powerful as editing your branch in a local clone of the repo and pushing your changes to your remote repo.

Hopefully, each PR has a **_"Allow edits from maintainers."_** checkbox, and if you did not uncheck it when you created the PR then any member with write access to the HaikuPorts repo can update your PR. I know how to do this and I do it from time to time, so I can update the changes to your PR if you wish/agree.

Of course, in an ideal world each contributor updates their own PRs so that the members only need to write reviews and check that the changes are ok.
But don't worry. I'll push the changes for you, including the ones spotted by @Begasus.

In the meanwhile you might wish to read some guides on GitHub :smile:

**Comment by alaviss** [comment]
@fbrosson 

> so I can update the changes to your PR if you wish/agree.

That's cheating. They have to learn git if they want to contribute.

@theclue You should modify the recipe as suggested by @fbrosson and @Begasus, then push the changes to your branch, which will updates this PR automatically :) Good luck. If you need help, feel free to ask on IRC ;)

**Comment by alaviss** [comment]
Originally I asked @theclue to port it as an exercise... I think it was best to keep this as an example recipe than being an actual port...

**Comment by fbrosson** [comment]
Sorry @alaviss :)
But, after all, having a recipe for GNU Hello is cool, isn't it? An OS which does not have GNU Hello cannot be considered as feature complete :smile:
BTW, another nice way to learn how to write recipe is to read (and build) existing ones.

**Comment by theclue** [comment]
sorry, didn't want to make troubled water here :) this is not that i don't want to RTFM of Github (i did actually, although the PR cycle is still a bit confusing but i'm working on it).

It is that I was missing about the best guidelines to follow (fixing and resubmitting the PR amending previous commits vs not putting the hands anymore into the code that was reviewed/validated by the mantainers and let them merge it in the best way/at the best time. Both are IMHO technically correct and which one to choose depends on the repo guidelines)

Next time I'll follow these :)

The recipe was intended as an exercise only, but I'm happy it was merged i must admit ;)

I was a bit disappointed by the fact that I was not able to indentify correctly the dependencies (following the docs) and that the autogenerated makefile is absolutely unreadable. Need to work on it, too.

I'm going to learn deeper the porting process. It's intriguing, although not so easy for a non-dev. I will improve with exercise

File: app-misc/hello/hello-2.10.recipe (hunk 6 above)
Comment (fbrosson) [reply app-misc/hello/hello-2.10.recipe]: > I've found about this dep into the README.dev. Is there a better way to check for dependencies (and the minimum version)?

I'd say that the README-dev is a good start. Then, if I suspect that a dependency is not really needed, I try to build the package without that dependency and check if it is still building fine and if the resulting package has the same contents. To compare the contents I use "package extract ..." and then make sure that every runtime (or library) has the same checksum.

**Comment by fbrosson** [comment]
There are 2 trends for updating a PR:

Method | Pros | Cons
-- | -- | --
Add commits | Easier to follow. | The member who merges the PR has to remember to edit the commit message (after clicking _"Squash and merge"_ and before _"Confirm merge"_) in order to remove the commit messages which are not relevant.
Replace the existing commit. | Easy to merge, and no risk of having garbage in the final commit message :smile: | The history of the changes is only visible if there were comments attached to the lines that have been changed. But that's not a big issue.

Both methods are OK, and I don't think one is better than the other.

**Comment by Begasus** [comment]
You could also create a sandbox branch and play around in there with pushing comments (for instance you could use <commit -m "message"> for every new commit, or <commit --amend> to squash the latest changes together with the previous one, no harm (if you keep away from your master branch) :)

**Pull request merged**